ir/function.cpp

```diff
@@ -428,5 +438,127 @@ void DomTree::printDot(std::ostream &os) const {
   os << "}\n";
 }
 
+
+void LoopAnalysis::getDepthFirstSpanningTree() {
+  std::map<const BasicBlock*, std::vector<const BasicBlock*>> edges;
+  unsigned current = 0;
+  for (auto [src, dst, instr] : cfg) {
+    edges[&src].push_back(&dst);
+  }
+
+  std::function<void(const BasicBlock *)> dfs = [&](const BasicBlock *bb) {
+    node.push_back(bb);
+    number[bb] = current++;
+    for (auto tgt : edges [bb]) {
+      if (number.find(tgt) == number.end()) {
+        dfs(tgt);
+      }
+    }
+    last[number[bb]] = current - 1;
+  };
+
+  dfs(&f.getFirstBB());
+}
+
+void LoopAnalysis::printNodeType(ostream &os, const NodeType &t) const {
+  switch (t) {
+  case nonheader:   os<<"n."; break;
+  case self:        os<<"s."; break;
+  case reducible:   os<<"r."; break;
+  case irreducible: os<<"i."; break;
+  }
+}
+
+// Implemention of Tarjan-Havlak algorithm.
+//
+// Irreducible loops are partially supported yet.
+//
+// Tarjan, R. (1974). Testing Flow Graph Reducibility.
+// Havlak, P. (1997). Nesting of reducible and irreducible loops.
+void LoopAnalysis::analysis() {
+  getDepthFirstSpanningTree();
+  unsigned bb_count = f.getBBs().size();
+
+  auto isAncestor = [this](unsigned w, unsigned v) -> bool {
+    return w <= v && v <= last.at(w);
+  };
+
+  vector<set<unsigned>> nonBackPreds, backPreds;
+
+  for (unsigned w = 0; w < bb_count; ++w) {
+    nonBackPreds.push_back({});
+    backPreds.push_back({});
+    header.push_back(0);
+    type.push_back(NodeType::nonheader);
+    for (auto [src, dst, instr] : cfg) {
```

Comment:
after all those vector initializations are moved out, this loop can be denested. Right now you have a quadratic loop here that can be avoided (I know it matches the paper, but since we don't have an efficient lookup from dst to src we should avoid being quadratic in the number of BBs).